File: LeetCode/Java/Solution/src/ios/ac/cn/ConstructBinaryTreefromPreorderandInorderTraversal.java
package ios.ac.cn;

public class ConstructBinaryTreefromPreorderandInorderTraversal {
	public TreeNode buildTree(int[] preorder, int[] inorder) {
		return buildTree(preorder, 0, preorder.length - 1, inorder, 0,
				inorder.length - 1);
	}
	
	private TreeNode buildTree(int[] preorder, int preLow, int preHi, int[] inorder,
			int inLow, int inHi) {
		if (preLow > preHi) {
			return null;
		} else {
			TreeNode root = new TreeNode(preorder[preLow]);
			if (preLow < preHi) {
				int rootIndex = inLow;
				for (int i = inLow; i<= inHi; i++) {
					if (inorder[i] == preorder[preLow]) {
						rootIndex = i;
						break;
					}
				}
				
				TreeNode left = buildTree(preorder, preLow + 1, preLow + (rootIndex - inLow ),
						inorder, inLow, rootIndex - 1);
				TreeNode right = buildTree(preorder, preLow + rootIndex - inLow + 1, preHi,
						inorder, rootIndex + 1, inHi);
				root.left = left;
				root.right = right;
			}
			return root;
		}
	}
}

File: LeetCode/Python/construct_binary_tree_from_preorder_and_inorder_traversal.py
#!/usr/bin/env python
#coding: utf-8

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param preorder, a list of integers
    # @param inorder, a list of integers
    # @return a tree node
    def buildTree(self, preorder, inorder):
        if not preorder: return None
        root = TreeNode(preorder[0])
        root_i = inorder.index(preorder[0])
        
        PL = preorder[1:root_i+1]
        PR = preorder[root_i+1:]
        IL = inorder[:root_i]
        IR = inorder[root_i + 1:]

        L = self.buildTree(PL, IL)
        R = self.buildTree(PR, IR)

        root.left = L
        root.right = R
        return root

if __name__ == '__main__':
    s = Solution()
    t = s.buildTree([1,2,3], [2,1,3])
    assert 1 == t.val    
    assert 2 == t.left.val
    assert 3 == t.right.val

